protocol.go

```diff
@@ -0,0 +1,426 @@
+func (p *Protocol) writeAcksIfNecessary() []byte {
+	for {
+		header, needed := p.createAckIfNecessary()
+		if !needed {
+			return nil
+		}
+
+		//log.Printf("%s: ack     (seq=%05d) (ack=%05d) (ack_bits=%032b)", p.conn.LocalAddr(), header.Sequence, header.ACK, header.ACKBits)
+
+		return p.write(header, nil)
+	}
+}
+
+func (p *Protocol) readAckBits(ack uint16, ackBits uint32) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for idx := uint16(0); idx < ACKBitsetSize; idx, ackBits = idx+1, ackBits>>1 {
+		if ackBits&1 == 0 {
+			continue
+		}
+
+		i := (ack - idx) % uint16(len(p.wq))
+		if p.wq[i] != uint32(ack-idx) || p.wqe[i].acked {
+			continue
+		}
+
+		if p.wqe[i].buf != nil {
+			p.pool.Put(p.wqe[i].buf)
+		}
+
+		p.wqe[i].buf = nil
+		p.wqe[i].acked = true
+	}
+}
+
+func (p *Protocol) trackRead(idx uint16) bool {
+	i := idx % uint16(len(p.rq))
+
+	if p.rq[i] == uint32(idx) { // duplicate packet
+		return false
+	}
+
+	if seq.GT(idx+1, p.ri) {
+		p.clearReads(p.ri, idx)
+		p.ri = idx + 1
+	}
+
+	p.rq[i] = uint32(idx)
+
+	return true
+}
+
+func (p *Protocol) clearReads(start, end uint16) {
+	count, size := end-start+1, uint16(len(p.rq))
+
+	if count >= size {
+		emptyBufferIndices(p.rq)
+		return
+	}
+
+	first := p.rq[start%size:]
+	length := uint16(len(first))
+
+	if count <= length {
+		emptyBufferIndices(first[:count])
+		return
+	}
+
+	second := p.rq[:count-length]
+
+	emptyBufferIndices(first)
+	emptyBufferIndices(second)
+}
+
+func (p *Protocol) trackAcked(ack uint16) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	lui := p.lui
+
+	for lui <= ack {
+		if p.rq[lui%uint16(len(p.rq))] != uint32(lui) {
+			break
+		}
+		lui++
+	}
+
+	p.lui = lui
+	p.ls = time.Now()
+}
+
+func (p *Protocol) trackUnacked() {
+	oui := p.oui
+
+	for {
+		i := oui % uint16(len(p.wq))
+		if p.wq[i] != uint32(oui) || !p.wqe[i].acked {
+			break
+		}
+		oui++
+	}
+	p.oui = oui
+
+	p.ouc.Broadcast()
+}
+
+func (p *Protocol) close() bool {
+	if p.die {
+		return false
+	}
+	close(p.exit)
+	p.die = true
+	p.ouc.Broadcast()
+
+	return true
+}
+
+func (p *Protocol) Close() {
+	if !p.close() {
+		return
+	}
+
+	//p.mu.Lock()
+	//defer p.mu.Unlock()
+
+	//if strings.Contains(p.conn.LocalAddr().String(), "44444") { // sending
+	//log.Printf("send closed (oldest_sent_ack_idx=%05d) (oldest_unacked_idx=%05d)", p.lui, p.oui)
+	//} else if strings.Contains(p.conn.LocalAddr().String(), "55555") { // receiving
+	//log.Printf("recv closed (oldest_sent_ack_idx=%05d) (oldest_unacked_idx=%05d)", p.lui, p.oui)
+	//}
+}
+
+func (p *Protocol) Run(transmit transmitFunc) {
+	ticker := time.NewTicker(p.updatePeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-p.exit:
+			return
+		case <-ticker.C:
+			if err := p.retransmitUnackedPackets(transmit); err != nil && p.eh != nil {
+				p.eh(p.addr, err)
+			}
+		}
+	}
+}
+
+func (p *Protocol) retransmitUnackedPackets(transmit transmitFunc) error {
+	for idx := uint16(0); idx < uint16(len(p.wq)); idx++ {
+		i := (p.oui + idx) % uint16(len(p.wq))
```

Comment:
Shouldn't there be a mutex lock here as well for `oui`, `wq`, etc?

Reply:
Yes, absolutely.